compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TaintAnalyzer.java

```diff
@@ -487,58 +496,64 @@ public void visit(BLangVariableDef varDefNode) {
+
+        Boolean combinedTaintedStatus = null;
+        if (varRefTaintedStatus != null && exprTaintedStatus != null) {
+            combinedTaintedStatus = varRefTaintedStatus || exprTaintedStatus;
+        }
         visitAssignment(compoundAssignment.varRef, combinedTaintedStatus, compoundAssignment.pos);
     }
 
-    private void visitAssignment(BLangExpression varRefExpr, boolean varTaintedStatus, DiagnosticPos pos) {
-        // Generate error if a global variable has been assigned with a tainted value.
-        if (varTaintedStatus && varRefExpr instanceof BLangVariableReference) {
-            BLangVariableReference varRef = (BLangVariableReference) varRefExpr;
-            if (varRef.symbol != null && varRef.symbol.owner != null) {
-                if (varRef.symbol.owner instanceof BPackageSymbol || SymbolKind.SERVICE.equals(varRef.symbol.owner.kind)
-                        || SymbolKind.CONNECTOR.equals(varRef.symbol.owner.kind)) {
-                    addTaintError(pos, varRef.symbol.name.value,
-                            DiagnosticCode.TAINTED_VALUE_PASSED_TO_GLOBAL_VARIABLE);
-                    return;
+    private void visitAssignment(BLangExpression varRefExpr, Boolean varTaintedStatus, DiagnosticPos pos) {
+        if (varTaintedStatus != null) {
+            // Generate error if a global variable has been assigned with a tainted value.
+            if (varTaintedStatus && varRefExpr instanceof BLangVariableReference) {
+                BLangVariableReference varRef = (BLangVariableReference) varRefExpr;
+                if (varRef.symbol != null && varRef.symbol.owner != null) {
+                    if (varRef.symbol.owner instanceof BPackageSymbol
+                            || SymbolKind.SERVICE.equals(varRef.symbol.owner.kind)
+                            || SymbolKind.CONNECTOR.equals(varRef.symbol.owner.kind)) {
+                        addTaintError(pos, varRef.symbol.name.value,
+                                DiagnosticCode.TAINTED_VALUE_PASSED_TO_GLOBAL_VARIABLE);
+                        return;
+                    }
                 }
             }
-        }
-        // TODO: Re-evaluating the full data-set (array) when a change occur.
-        if (varRefExpr instanceof BLangIndexBasedAccess) {
-            nonOverridingAnalysis = true;
-            updatedVarRefTaintedState((BLangIndexBasedAccess) varRefExpr, varTaintedStatus);
-            nonOverridingAnalysis = false;
-        } else if (varRefExpr instanceof BLangFieldBasedAccess) {
-            BLangFieldBasedAccess fieldBasedAccessExpr = (BLangFieldBasedAccess) varRefExpr;
-            // Propagate tainted status to fields, when field symbols are present (Example: struct).
-            if (fieldBasedAccessExpr.symbol != null) {
-                setTaintedStatus(fieldBasedAccessExpr, varTaintedStatus);
+            // TODO: Re-evaluating the full data-set (array) when a change occur.
+            if (varRefExpr instanceof BLangIndexBasedAccess) {
```

Comment:
Can we check NodeKind here instead of instanceof?

Reply:
Yes +1